pkg/copier/copier.go

```diff
@@ -0,0 +1,94 @@
+package copier
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"time"
+
+	"github.com/block/spirit/pkg/dbconn"
+	"github.com/block/spirit/pkg/metrics"
+	"github.com/block/spirit/pkg/table"
+	"github.com/block/spirit/pkg/throttler"
+	"github.com/siddontang/go-log/loggers"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	copyEstimateInterval   = 10 * time.Second // how frequently to re-estimate copy speed
+	copyETAInitialWaitTime = 1 * time.Minute  // how long to wait before first estimating copy speed (to allow for fast start)
+)
+
+// Copier is the interface which copiers use. Currently we only have
+// one implementation, which we call unbuffered because it uses
+// INSERT .. SELECT without any intermediate buffering in spirit.
+// In future we may have another implementation, see:
+// https://github.com/block/spirit/issues/451
+//
+// Yes, this interface is a bit bloated. We will have to remove the
+// deprecated methods, but that's for the future.
+type Copier interface { //nolint: interfacebloat
+	Run(ctx context.Context) error
+	GetETA() string
+	GetChunker() table.Chunker
+	SetThrottler(throttler throttler.Throttler)
+	GetThrottler() throttler.Throttler
+	StartTime() time.Time
+	GetProgress() string
+
+	// The following are for testing purposes only
```

Comment:
I think these could be unexported if they're only used for tests in this package *and* if you don't ever intend to implement this interface in another package.